src/api.rs

```diff
@@ -1,33 +1,136 @@
 // API endpoints
 
-use actix_web::{post, get, Responder, web::Payload};
+use crate::game::adapter::GameAdapter;
+use crate::game::{adapter, connect4, GameId, GameManager};
+use actix_web::http::StatusCode;
+use actix_web::web::Json;
+use actix_web::{get, post, web, Error, Responder, ResponseError, Result};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::fmt::{write, Display, Formatter};
+use std::ops::DerefMut;
+use std::process::id;
+use std::sync::RwLock;
+use crate::game::ValidationError::NoSuchGameError;
+
+/* Helpers */
+
+/* Structs for the different endpoints*/
+
+/// Struct for req to `/api/create-game`
+#[derive(Deserialize)]
+pub struct CreateGameReq {
+    name: String,
+}
+
+/// Struct for res from `/api/create-game`
+#[derive(Serialize)]
+pub struct CreateGameRes {
+    game_id: String,
+}
+
+/// Struct for res from `/api/list-games`
+#[derive(serde::Serialize)]
+pub struct ListGamesRes {
+    // TODO
+}
+
+/// Struct for req to `/api/join-game`
+#[derive(Deserialize)]
+pub struct JoinGameReq {
+    game_id: String,
+    username: String,
+}
+
+/// Struct for res from `/api/join-game`
+#[derive(serde::Serialize)]
+pub struct JoinGameRes {
+    session_id: String,
+}
+
+// Struct for res from `/api/{game_id}/get-state`
+// <br> It's basically the game::adaptor::GenericGameState Object
+
+// Struct for req to `/api/{game_id}/submit-move`
+// This is basically the game::adaptor::GenericGameMove struct
+
+/// Struct for res from `/api/{game_id}/submit-move`
+#[derive(Serialize)]
+pub struct SubmitMoveRes {
+    success: bool,
+}
+
+/// Struct for res from `/api/{game_id}/submit-move`
+#[derive(Serialize)]
+pub struct WaitForUpdateRes {
+    updated: bool,
+}
 
 #[post("/api/create-game")]
-pub(crate) async fn create_game(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn create_game(
+    mut payload: web::Json<CreateGameReq>,
+    gm_wrapped: web::Data<RwLock<GameManager>>,
+) -> Result<Json<CreateGameRes>> {
+    // Validate & Pass it on
+    let mut gm = gm_wrapped.write().unwrap();
+    let game_id = match payload.name.as_str() {
+        "connect_4" => gm.create_game(|id| Box::new(connect4::Connect4Adapter::new(id))),
+        _ => return Err(Error::from(NoSuchGameError(payload.name.clone()))),
+    };
+    game_id.and_then(|id| {
+        Ok(Json(CreateGameRes {
+            game_id: id.to_string()[4..].parse()?,
```

Comment:
This is very suspicious... if the `Display` implementation for `GameId` returns something like "Game12345", why do we strip the "Game" here before sending it to the client? Wouldn't it be helpful when debugging for a user to know at first glance what sort of id they have?